url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/block/BlockMatrixOps.java

```diff
@@ -0,0 +1,674 @@
+/*
+ * Copyright (c) 2009-2013, Peter Abeles. All Rights Reserved.
+ *
+ * This file is part of Efficient Java Matrix Library (EJML).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ejml.alg.block;
+
+import org.ejml.data.BlockMatrix64F;
+import org.ejml.data.D1Submatrix64F;
+import org.ejml.data.DenseMatrix64F;
+import org.ejml.ops.CommonOps;
+import org.ejml.ops.MatrixFeatures;
+import org.ejml.ops.RandomMatrices;
+import org.ejml.simple.SimpleMatrix;
+
+import java.util.Random;
+
+
+/**
+ * Various operations on {@link BlockMatrix64F}.
+ *
+ * @author Peter Abeles
+ */
+public class BlockMatrixOps {
+
+    /**
+     * Converts a row major matrix into a row major block matrix.
+     *
+     * @param src Original DenseMatrix64F.  Not modified.
+     * @param dst Equivalent BlockMatrix64F. Modified.
+     */
+    public static void convert( DenseMatrix64F src , BlockMatrix64F dst )
+    {
+        if( src.numRows != dst.numRows || src.numCols != dst.numCols )
+            throw new IllegalArgumentException("Must be the same size.");
+
+        for( int i = 0; i < dst.numRows; i += dst.blockLength ) {
+            int blockHeight = Math.min( dst.blockLength , dst.numRows - i);
+
+            for( int j = 0; j < dst.numCols; j += dst.blockLength ) {
+                int blockWidth = Math.min( dst.blockLength , dst.numCols - j);
+
+                int indexDst = i*dst.numCols + blockHeight*j;
+                int indexSrcRow = i*dst.numCols + j;
+
+                for( int k = 0; k < blockHeight; k++ ) {
+                    System.arraycopy(src.data,indexSrcRow,dst.data,indexDst,blockWidth);
+                    indexDst += blockWidth;
+                    indexSrcRow += dst.numCols;
+                }
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 50:65, 123:138.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.